src/components/TransactionIOList.tsx

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { AssetOutput, Input, Output } from '@alephium/sdk/api/explorer'
+import { ALPH } from '@alephium/token-list'
+import { ReactElement, ReactNode } from 'react'
+
+import { AddressLink } from './Links'
+
+interface TransactionIOListProps {
+  ioList: Input[] | Output[]
+  flex?: boolean
+  addressMaxWidth?: string
+  IOItemWrapper?: ({ children }: { children: ReactNode }) => ReactElement
+}
+
+const TransactionIOList = ({ ioList, flex, addressMaxWidth, IOItemWrapper }: TransactionIOListProps) => (
+  <>
+    {ioList.map((io, i) => {
+      if (!io.address) return null
+
+      const amounts = [{ id: ALPH.id, amount: BigInt(io.attoAlphAmount ?? 0) }]
+
+      if (io.tokens) {
+        amounts.push(...io.tokens.map((t) => ({ id: t.id, amount: BigInt(t.amount) })))
+      }
+
+      const IOAddressLink = (
+        <AddressLink
+          address={io.address}
+          txHashRef={(io as Input).txHashRef}
+          lockTime={(io as AssetOutput).lockTime}
```

Comment:
this looks like a hack :/ The `io` cannot be both `Input` and `AssetOutput` at the same time, right?

I took the liberty to play a bit and find a different solution that only requires casting an `Output` as an `AssetOutput` (which is better than the current solution imo).

What do you think about the following implementation?

No need to spend too much time on this, I though this is an easy win for a cleaner code-base.

```tsx
interface TransactionIOListProps {
  inputs?: Input[]
  outputs?: Output[]
  flex?: boolean
  addressMaxWidth?: string
  IOItemWrapper?: ({ children }: { children: ReactNode }) => ReactElement
}

const TransactionIOList = ({
  inputs = [],
  outputs = [],
  flex,
  addressMaxWidth,
  IOItemWrapper
}: TransactionIOListProps) => {
  const getAmounts = (io: Input | Output) => [
    { id: ALPH.id, amount: BigInt(io.attoAlphAmount ?? 0) },
    ...(io.tokens ? io.tokens.map((t) => ({ id: t.id, amount: BigInt(t.amount) })) : [])
  ]

  const renderLink = (IOAddressLink: ReactNode, addressHash: string, index: number) =>
    IOItemWrapper !== undefined ? (
      <IOItemWrapper key={`${addressHash}-${index}`}>{IOAddressLink}</IOItemWrapper>
    ) : (
      IOAddressLink
    )

  return (
    <>
      {inputs.map((input, i) =>
        !input.address
          ? null
          : renderLink(
              <AddressLink
                address={input.address}
                txHashRef={input.txHashRef}
                amounts={getAmounts(input)}
                maxWidth={addressMaxWidth}
                flex={flex}
              />,
              input.address,
              i
            )
      )}
      {outputs.map((output, i) =>
        !output.address
          ? null
          : renderLink(
              <AddressLink
                address={output.address}
                lockTime={(output as AssetOutput).lockTime}
                amounts={getAmounts(output)}
                maxWidth={addressMaxWidth}
                flex={flex}
              />,
              output.address,
              i
            )
      )}
    </>
  )
}

export default TransactionIOList
```

Reply:
> The io cannot be both Input and AssetOutput at the same time, right?

Exactly, we know that for sure :) That's why I found my solution rather elegant. Taking a look a yours 👁️ ...

`Reading...`

I liked my solution TBH, mostly because it's DRYier and shorter. But I see your point, and I'm fine with your solution as well! 👍  Updated.